File: PHP/One/Transforms/InferTypesPlugins/LambdaResolver.php
<?php

namespace One\Transforms\InferTypesPlugins\LambdaResolver;

use One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use One\Ast\Expressions\Expression;
use One\Ast\Types\Lambda;
use One\Ast\AstTypes\LambdaType;
use One\Ast\AstTypes\TypeHelper;

class LambdaResolver extends InferTypesPlugin {
    function __construct() {
        parent::__construct("LambdaResolver");
        
    }
    
    protected function setupLambdaParameterTypes($lambda) {
        if ($lambda->expectedType === null)
            return;
        
        if ($lambda->expectedType instanceof LambdaType) {
            $declParams = $lambda->expectedType->parameters;
            if (count($declParams) !== count($lambda->parameters))
                $this->errorMan->throw("Expected " . count($lambda->parameters) . " parameters for lambda, but got " . count($declParams) . "!");
            else
                for ($i = 0; $i < count($declParams); $i++) {
                    if ($lambda->parameters[$i]->type === null)
                        $lambda->parameters[$i]->type = $declParams[$i]->type;
                    else if (!TypeHelper::isAssignableTo($lambda->parameters[$i]->type, $declParams[$i]->type))
                        $this->errorMan->throw("Parameter type " . $lambda->parameters[$i]->type->repr() . " cannot be assigned to " . $declParams[$i]->type->repr() . ".");
                }
        }
        else
            $this->errorMan->throw("Expected LambdaType as Lambda's type!");
    }
    
    protected function visitLambda($lambda) {
        $this->setupLambdaParameterTypes($lambda);
    }
    
    function canTransform($expr) {
        return $expr instanceof Lambda;
    }
    
    function transform($expr) {
        $this->visitLambda($expr);
        // does not transform actually
        return $expr;
    }
}

File: CSharp/One/Transforms/InferTypesPlugins/LambdaResolver.cs
using One.Ast;
using One.Transforms.InferTypesPlugins.Helpers;

namespace One.Transforms.InferTypesPlugins
{
    public class LambdaResolver : InferTypesPlugin
    {
        public LambdaResolver(): base("LambdaResolver")
        {
            
        }
        
        protected void setupLambdaParameterTypes(Lambda lambda)
        {
            if (lambda.expectedType == null)
                return;
            
            if (lambda.expectedType is LambdaType lambdType) {
                var declParams = lambdType.parameters;
                if (declParams.length() != lambda.parameters.length())
                    this.errorMan.throw_($"Expected {lambda.parameters.length()} parameters for lambda, but got {declParams.length()}!");
                else
                    for (int i = 0; i < declParams.length(); i++) {
                        if (lambda.parameters.get(i).type == null)
                            lambda.parameters.get(i).type = declParams.get(i).type;
                        else if (!TypeHelper.isAssignableTo(lambda.parameters.get(i).type, declParams.get(i).type))
                            this.errorMan.throw_($"Parameter type {lambda.parameters.get(i).type.repr()} cannot be assigned to {declParams.get(i).type.repr()}.");
                    }
            }
            else
                this.errorMan.throw_("Expected LambdaType as Lambda's type!");
        }
        
        protected void visitLambda(Lambda lambda)
        {
            this.setupLambdaParameterTypes(lambda);
        }
        
        public override bool canTransform(Expression expr)
        {
            return expr is Lambda;
        }
        
        public override Expression transform(Expression expr)
        {
            this.visitLambda(((Lambda)expr));
            // does not transform actually
            return expr;
        }
    }
}